pkg/query/statistics_optimizer.go

```diff
@@ -0,0 +1,268 @@
+package query
+
+import (
+	"cmp"
+	"slices"
+)
+
+// StatisticsOptimizer uses cost estimates to optimize iterator trees.
+type StatisticsOptimizer struct {
+	Source StatisticsSource
+}
+
+// Optimize applies statistics-based optimizations to an iterator tree.
+// It uses cost estimates to make decisions about reordering and restructuring.
+func (s StatisticsOptimizer) Optimize(it Iterator) (Iterator, bool, error) {
+	// Apply all statistics-based optimization strategies
+	optimizers := []OptimizerFunc{
+		s.reorderUnion,
+		s.reorderIntersection,
+		s.rebalanceArrow,
+	}
+
+	return ApplyOptimizations(it, optimizers)
+}
+
+// reorderUnion reorders union subiterators by selectivity (higher first).
+// Higher selectivity branches are more likely to short-circuit, making unions more efficient.
+func (s StatisticsOptimizer) reorderUnion(it Iterator) (Iterator, bool, error) {
+	union, ok := it.(*Union)
+	if !ok {
+		return it, false, nil
+	}
+
+	newSubs, changed, err := s.reorderBySelectivity(union.subIts, false)
+	if err != nil || !changed {
+		return it, changed, err
+	}
+
+	return NewUnion(newSubs...), true, nil
```

Comment:
Don't we usually parallelize these dispatches? Does reordering still work in that case? Or are we leaving that as a later exercise?